src/processor/operator/recursive_extend/path_scanner.cpp

```diff
@@ -0,0 +1,102 @@
+#include "processor/operator/recursive_extend/path_scanner.h"
+
+namespace kuzu {
+namespace processor {
+
+size_t FixedLengthPathScanner::scanPaths(common::table_id_t tableID,
+    common::ValueVector* pathOffsetVector, common::ValueVector* pathDataVector,
+    common::ValueVector* dstNodeIDVector, common::sel_t& offsetVectorPos,
+    common::sel_t& dataVectorPos) {
+    if (numRels >= bfsMorsel.frontiers.size()) {
+        // BFS terminate before current depth. No need to scan.
+        return 0;
+    }
+    auto offsetVectorPosBeforeScanning = offsetVectorPos;
+    auto lastFrontier = bfsMorsel.frontiers[numRels].get();
+    while (offsetVectorPos < common::DEFAULT_VECTOR_CAPACITY) {
+        if (!continueDfs()) {
+            // All paths from current dst have been scanned. Fetch next dst.
+            if (lastFrontierCursor == lastFrontier->nodeOffsets.size()) {
+                // All nodes from last frontier have been scanned.
+                break;
+            }
+            auto currentDstOffset = lastFrontier->nodeOffsets[lastFrontierCursor++];
+            // Skip nodes that is not in semi mask.
+            if (!bfsMorsel.isAllDstTarget() &&
+                !bfsMorsel.targetDstNodeOffsets.contains(currentDstOffset)) {
+                continue;
+            }
+            // Prepare dfs stack.
+            initDfs(currentDstOffset, numRels);
+        }
+        writePathToVector(tableID, pathOffsetVector, pathDataVector, dstNodeIDVector,
+            offsetVectorPos, dataVectorPos);
+    }
+    return offsetVectorPos - offsetVectorPosBeforeScanning;
+}
+
+bool FixedLengthPathScanner::continueDfs() {
+    auto level = 0;
+    while (!nbrsStack.empty()) {
+        auto& cursor = cursorStack.top();
+        cursor++;
+        if (cursor < nbrsStack.top()->size()) { // Found a new nbr
+            auto offset = nbrsStack.top()->at(cursor);
+            path[level] = offset;
+            if (level == 0) { // Found a new nbr at level 0. Found a new path.
+                return true;
```

Comment:
I would do this:

```
void FixedLengthPathScanner::continueDfs(common::table_id_t tableID,
    common::ValueVector* pathOffsetVector, common::ValueVector* pathDataVector,
    common::ValueVector* dstNodeIDVector, common::sel_t& offsetVectorPos,
    common::sel_t& dataVectorPos) {
    auto level = 0;
    while (!nbrsStack.empty()) {
        auto& cursor = cursorStack.top();
        cursor++;
        if (cursor < nbrsStack.top()->size()) { // Found a new nbr
            auto offset = nbrsStack.top()->at(cursor);
            path[level] = offset;
            if (level == 0) { // Found a new nbr at level 0. Found a new path.
               writePathToVector(tableID, pathOffsetVector, pathDataVector, dstNodeIDVector,
                    offsetVectorPos, dataVectorPos);
               if (offsetVectorPos == common::DEFAULT_VECTOR_CAPACITY) {
                    return;
               } 
           }
            // Push new stack.
            cursorStack.push(-1);
            nbrsStack.push(&bfsMorsel.frontiers[level]->bwdEdges.at(offset));
            level--;
        } else { // Failed to find a nbr. Pop stack.
            cursorStack.pop();
            nbrsStack.pop();
            level++;
        }
    }
}
```

The number of continueDFS() function calls you are making right now is commensurate with the number and lengths of the paths. You can avoid these. You can also make `writePathToVector` an inlined function. The modified function I wrote about writes as many paths as it can, up to common::DEFAULT_VECTOR_CAPACITY many paths. Then the caller can check if it needs to start a new DFS  or not. If you want you can make this function return true to indicate you have written as many as common::DEFAULT_VECTOR_CAPACITY and false if there are more paths to write, which the caller can use to decide whether to start a new DFS or not.